jib-core/src/main/java/com/google/cloud/tools/jib/api/RegistryImage.java

```diff
@@ -0,0 +1,125 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.api;
+// TODO: Move to com.google.cloud.tools.jib once that package is cleaned up.
+
+import com.google.cloud.tools.jib.configuration.credentials.Credential;
+import com.google.cloud.tools.jib.configuration.credentials.CredentialRetriever;
+import com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory;
+import com.google.cloud.tools.jib.image.ImageReference;
+import com.google.cloud.tools.jib.image.InvalidImageReferenceException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Builds to a container registry.
+ *
+ * <p>The registry portion of the image reference determines which registry to push the image to.
+ * The repository portion is the namespace to push the image to. The tag is a label to easily
+ * identify an image among all the images in the repository. See {@link ImageReference} for more
+ * details.
+ *
+ * <p>When configuring credentials (via {@link #setCredential} for example), make sure the
+ * credentials are valid push credentials for the repository specified via the image reference.
+ */
+// TODO: Add tests once JibContainerBuilder#containerize() is added.
+public class RegistryImage {
+
+  /**
+   * Instantiate with the image reference to push to.
+   *
+   * @param imageReference the image reference
+   * @return a new {@link RegistryImage}
+   */
+  public static RegistryImage named(ImageReference imageReference) {
+    return new RegistryImage(imageReference);
+  }
+
+  /**
+   * Instantiate with the image reference to push to.
+   *
+   * @param imageReference the image reference
+   * @return a new {@link RegistryImage}
+   * @throws InvalidImageReferenceException if {@code imageReference} is not a valid image reference
+   */
+  public static RegistryImage named(String imageReference) throws InvalidImageReferenceException {
+    return named(ImageReference.parse(imageReference));
+  }
+
+  private final ImageReference imageReference;
+  private List<CredentialRetriever> credentialRetrievers = new ArrayList<>();
+
+  /** Instantiate with {@link #named}. */
+  private RegistryImage(ImageReference imageReference) {
+    this.imageReference = imageReference;
+  }
+
+  /**
+   * Sets the credentials to use to push the image. This replaces any previously-added {@link
+   * CredentialRetriever}s.
+   *
+   * @param username the username
+   * @param password the password
+   * @return this
+   */
+  public RegistryImage setCredential(String username, String password) {
+    credentialRetrievers = new ArrayList<>();
```

Comment:
It feels weird to have a `setCredential` that removes all other credentials in a builder. Would people have any reason to want to remove all credentials they've set, and if they do, would it be better to just explicitly have a `clearCredentials` method instead of overwriting with a new credential? Alternatively, should we have a `setCredential` that takes a generic `CredentialRetriever` instead of only allowing a username and password?

Reply:
Hmm, that is a good point that having `setCredential` mean clear credential retrievers and use these credentials instead might be putting too much on one setter. Perhaps we should name this `useCredential`? The alternative would be to add a `CredentialRetriever` that returns the provided credential to the list of credential retrievers, but that might be confusing since there's already an `addCredentialRetriever`. We could have a `setCredentialRetriever` that replaces the list of credential retrievers with a single credential retriever, but I think `setCredentialRetrievers` to set a list would make more sense in that case. However, I steered away from setting full lists of credential retrievers that are built somewhere else since a list of credential retrievers is very specific to the `RegistryImage` they are for. Also, having this `(String username, String password)` signature is mostly meant to be a shorthand for users to easily provide a basic credential without needing to mess with `CredentialRetriever` or `Credential` classes.

So I think the options I'm considering are:
1. Rename to `useCredential`
1. Rename to `withCredential`
1. Rename to `addCredential` and change to add instead of replace

Reply:
To me, option 3 makes the most sense; I can't think of a reason someone would want to add credentials and replace them later for the same registry.

Reply:
Yeah, there's a little bit of a disconnect between a single item and a list here. I would think `set...` mostly matches the underlying type (list). And `add` could augment the list.  I don't predict a lot of use from something like `clear` or `remove`.

Reply:
Assuming the tool developer has to know when to use basic credentials, I don't see it being particularly expensive for a tool developer to use the slightly longer form `addCrendentials(Credentials.basic(asdf,asfd))` or whatever.